common/migration_mechanism.py

```diff
@@ -0,0 +1,192 @@
+#!/usr/bin/env python3
+#
+# Copyright (C) 2018  Joseph Spearritt <mcoot@tamods.org>
+#
+# This file is part of taserver
+#
+# taserver is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as
+# published by the Free Software Foundation, either version 3 of the
+# License, or (at your option) any later version.
+#
+# taserver is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with taserver.  If not, see <http://www.gnu.org/licenses/>.
+#
+
+from typing import Dict, List
+
+from collections import OrderedDict
+from functools import wraps
+import os
+import shutil
+import json
+import copy
+import glob
+
+
+# Known migrations
+_registered_migrations = OrderedDict()
+
+
+def _perform_backups(data_root: str, schema_version: int) -> None:
+    backup_path = data_root + ('.%d.bkp' % schema_version)
+    if os.path.exists(backup_path):
+        shutil.rmtree(backup_path)
+    shutil.copytree(data_root, backup_path)
+
+
+def _load_schema_version(data_root: str) -> int:
+    if not os.path.exists(os.path.join(data_root, 'metadata.json')):
+        return 0
+    with open(os.path.join(data_root, 'metadata.json'), 'rt') as f:
+        metadata: Dict = json.load(f)
+        return metadata.get('schema_version', 0)
+
+
+def _save_schema_version(data_root: str, schema_version: int) -> None:
+    metadata = dict()
+    if os.path.exists(os.path.join(data_root, 'metadata.json')):
+        with open(os.path.join(data_root, 'metadata.json'), 'rt') as f:
+            metadata = json.load(f)
+    metadata['schema_version'] = schema_version
+    with open(os.path.join(data_root, 'metadata.json'), 'wt') as f:
+        json.dump(metadata, f, indent=4, sort_keys=True)
+
+
+def _get_players_to_migrate(data_root: str) -> List[str]:
+    """
+    Get the list of players currently registered
+
+    :param data_root: the root of the data store directory
+    :return: the list of registered player login names
+    """
+    if not os.path.exists(os.path.join(data_root, 'accountdatabase.json')):
+        # No account database
+        return []
+    with open(os.path.join(data_root, 'accountdatabase.json'), 'rt') as f:
+        accounts = json.load(f)
+    return [acc['login_name'] for acc in accounts if 'login_name' in acc]
```

Comment:
'login_name' is mandatory, right? No need to check for it.

Reply:
Fixed